linera-witty-macros/src/wit_interface.rs

```diff
@@ -0,0 +1,215 @@
+        inputs: impl IntoIterator<Item = &'input FnArg>,
+        output: ReturnType,
+    ) -> Self {
+        let (parameter_names, parameter_types) = inputs
+            .into_iter()
+            .map(|argument| {
+                let FnArg::Typed(PatType { pat, ty, .. }) = argument else {
+                    abort!(argument, "`self` is not supported in imported functions");
+                };
+
+                let Pat::Ident(PatIdent { ident, .. }) = pat.as_ref() else {
+                    abort!(
+                        pat,
+                        "Only named parameters are supported in imported functions"
+                    );
+                };
+
+                (ident, ty.as_ref())
+            })
+            .unzip();
+
+        let output_type = match output {
+            ReturnType::Default => None,
+            ReturnType::Type(_arrow, return_type) => Some(*return_type),
+        };
+
+        FunctionInformation {
+            name,
+            parameter_names,
+            parameter_types,
+            output_type,
+        }
+    }
+
+    /// Returns the types used in the function signature.
+    pub fn dependencies(&self) -> impl Iterator<Item = &'_ Type> {
+        self.parameter_types
+            .clone()
+            .into_iter()
+            .chain(&self.output_type)
+    }
+
+    /// Returns a [`LitStr`] with the kebab-case WIT name of the function.
+    pub fn wit_name(&self) -> LitStr {
+        LitStr::new(&self.name.to_string().to_kebab_case(), self.name.span())
+    }
+
+    /// Returns the code to generate a [`String`] with the WIT declaration of the function.
+    pub fn wit_declaration(&self) -> TokenStream {
+        let wit_name = self.wit_name();
+
+        let parameters = self.parameter_names.iter().zip(&self.parameter_types).map(
+            |(parameter_name, parameter_type)| {
+                let parameter_wit_name = LitStr::new(
+                    &parameter_name.to_string().to_kebab_case(),
+                    parameter_name.span(),
+                );
+
+                quote! {
+                    #parameter_wit_name.into(),
+                    ": ".into(),
+                    <#parameter_type as linera_witty::WitType>::wit_type_name()
+                }
+            },
+        );
+
+        let commas = iter::repeat(Some(quote! { ", ".into(), }))
+            .take(self.parameter_names.len().saturating_sub(1))
+            .chain(Some(None));
+
+        let output = self
+            .output_type
+            .as_ref()
+            .map(|output_type| {
+                quote! { " -> ".into(), <#output_type as linera_witty::WitType>::wit_type_name() }
+            })
+            .into_iter();
+
+        quote! {
+            [
+                std::borrow::Cow::Borrowed("    "),
+                #wit_name.into(),
+                ": func(".into(),
+                #( #parameters, #commas )*
+                ")".into(),
+                #( #output, )*
+                ";".into(),
```

Comment:
I'm still confused: Somehow I expected this would be a string literal, but it doesn't create any quotes (`"`)?

Reply:
I'm not sure I understand the question :sweat_smile: This creates a `String` object, using the [`slice::join` method](https://doc.rust-lang.org/std/primitive.slice.html#method.join) applied to [strings](https://doc.rust-lang.org/src/alloc/str.rs.html#73). This is very similar to `concat!`, but some [`WitType`s might need to build `String` objects](https://github.com/linera-io/linera-protocol/blob/abd870526dbb7c18abbab051040fefce14f68378/linera-witty/src/type_traits/mod.rs#L27-L31), so this concatenates a list of `Cow<'static, str>`.

Reply:
Ah, right, and `wit_name` and `parameters` _are_ string literals, of course. :+1: